### PR TITLE
Operators written purely in Jax

### PR DESCRIPTION
**Pull request by inailuig**
Currently the operators are written using numba  for the very valid reason that the array shapes (number of connected elements) are in general data-dependent which is something not (yet?) supported by jax.
These operators can only be evaluated on the cpu, and are called either outside of a jax jit block or via numba4jax which has some overhead.

This PR adds a Ising operator with padding written in jax and a branchless, jax-jittable MetropolisSampler (I wanted some way to test the operator...) that runs fully on the gpu.

For Paulistrings the same can also be done easily (apart from the constructor).

### CODE REVIEW

File: netket/sampler/rules/hamiltonian.py [hunk 1]
```diff
@@ -114,3 +116,70 @@ def _choose(vp, sections, rand_vec, out, w):
...
+    In this case, the transition matrix is taken to be:
+
+    .. math::
+
+       T( \\mathbf{s} \\rightarrow \\mathbf{s}^\\prime) = \\frac{1}{\\mathcal{N}(\\mathbf{s})}\\theta(|H_{\\mathbf{s},\\mathbf{s}^\\prime}|),
+
+    This rule only works with operators which are written in jax.
+    """
+
+    def transition(self, _0, _1, _2, _3, key, x):
```
Comment (gcarleo) [inline netket/sampler/rules/hamiltonian.py:135]: what are these 4 dummy arguments?

Comment (inailuig) [inline netket/sampler/rules/hamiltonian.py:135]:
```suggestion
    def transition(self, sampler, machine, parameters, state, key, x):
```

File: netket/operator/_discrete_operator.py [hunk 2]
```diff
@@ -40,6 +40,9 @@ def get_conn_padded(self, x: np.ndarray) -> Tuple[np.ndarray, np.ndarray]:
             mels: A N-tensor containing the matrix elements :math:`O(x,x')`
                 associated to each x' for every batch.
         """
+
+        x = np.asarray(x)
+
```
Comment (PhilipVinc) [inline netket/operator/_discrete_operator.py:45]: This change I'd like to see it in a separate PR, if possible, to keep noise down in this one to a minimum.

Can you or @inailuig open a PR inserting this change and removing all calls to asarray In mc_expect and co?

Please add a test as well

Comment (PhilipVinc) [reply netket/operator/_discrete_operator.py:45]: cc @wdphy16

Comment (wdphy16) [reply netket/operator/_discrete_operator.py:45]: Ok, I'll do it

Comment (PhilipVinc) [reply netket/operator/_discrete_operator.py:45]: actually I think it already landed in master.

Comment (wdphy16) [reply netket/operator/_discrete_operator.py:45]: Yes indeed

**Comment by inailuig** [comment]
@PhilipVinc This PR still contains a hamiltonian sampler rule for the jax operators.
Once we have merged that we can close it.
